### ISSUE TITLE
[Feature] Multi-Machine Support for Distributed Inference

### ISSUE THREAD

**Issue by ronaldmannak** [ml-explore/mlx#1046]
The growth in size of open-source models is outpacing the growth of memory capacity of Mac computers. The latest 70B version of Llama 3 is already pushing the limits of a fully loaded Mac Pro. The upcoming 400B version of Llama 3 will exceed available memory entirely unless heavily quantized.

While memory limits may increase in future Mac Pro and Mac Studio models, it is likely that LLMs will continue to grow in size at an even faster rate. This poses a challenge for running the latest large open-source models with MLX. Without changes, MLX could be restricted to handling small to medium-sized models or heavily quantized versions of large models, resulting in inevitable inaccuracies.

MLX may become unsuitable for scenarios where local GPT-4-equivalent open-source models are necessary for cost and/or privacy considerations. I'm particularly thinking of SMB businesses and power users.

If we discard lossy options like quantizing, there are alternative approaches to consider:

1) Optimizing memory usage. For instance, [Air_LLM](https://github.com/lyogavin/Anima/tree/main/air_llm), which involves loading and unloading layers. It is unclear if every LLM supports unloading entire layers, and this method may be inefficient since the layers have to be cycled for each generated token.

2) Implementing multi-machine support for distributed inference, where inference is distributed across multiple Macs. I shared a [tweet](https://x.com/ronaldmannak/status/1784087769817756138) about this possible solution and received significant interest, even though it was just a spontaneous idea. One way this approach could function is by splitting the model across multiple Macs (Mini, Studio, or Pro) connected via IP over Thunderbolt.

I am not proposing a definitive solution, but if there is interest in this topic, this discussion could serve as a starting point for further exploration of the possibilities.

**Comment by sck-at-ucy** [comment]
"Implementing multi-machine support for distributed inference": I am not sure how this would work in technical detail but this is an idea that would be _very_ attractive in the context of using MLX for physics/engineering ML/AI+HPC applications.

**Comment by ivanfioravanti** [comment]
I think we should extend the concept to distributed training and inference, like the deepseed library. This can enable incredible scenarios powered by Apple Silicons chips.

**Comment by awni** [comment]
As a first step we are looking at adding the communication primitives you would use to implement both of these. Ops like `send`, `receive`, `broadcast`, `reduce` etc. Basically the ops in this [table](https://pytorch.org/docs/stable/distributed.html#backends) / MPI.

Both distributed training and inference should be implementable on top of those ops. Exactly what those APIs look like and where they live is still TBD. But we need those ops as a first step either way so we can do distributed work with MLX in a flexible way.

**Comment by ronaldmannak** [comment]
@awni That sounds like a good and doable first step. If I have time this week, I'm happy to take a first stab at it.

FWIW, there is a relevant discussion about how the accuracy of Llama 3 model may be disproportionately affected by quantization, which might be a side effect of using a large training set. This would be another argument for distributed inference. See 
[tweet](https://x.com/rohanpaul_ai/status/1784972618472317180?s=46&t=CSJUc-EX4J03ow_Gi8B9Lw) and [PR](https://github.com/ggerganov/llama.cpp/pull/6936).

**Comment by fblissjr** [comment]
This is something I'd be happy to test and contribute to as well. I remember seeing the original tweet, and just now got the Thunderbolt 4 cable connected across two macs (macbook pro and a m2 ultra). 

Can see this use case being common for Apple users with a work and home machine.

**Comment by fblissjr** [comment]
> @awni That sounds like a good and doable first step. If I have time this week, I'm happy to take a first stab at it.
> 
> FWIW, there is a relevant discussion about how the accuracy of Llama 3 model may be disproportionately affected by quantization, which might be a side effect of using a large training set. This would be another argument for distributed inference. See [tweet](https://x.com/rohanpaul_ai/status/1784972618472317180?s=46&t=CSJUc-EX4J03ow_Gi8B9Lw) and [PR](https://github.com/ggerganov/llama.cpp/pull/6936).

From what i've read, this is a llama.cpp issue, more than just llama3 quants.

**Comment by awni** [comment]
In progress #1097

**Comment by fblissjr** [comment]
> In progress #1097

amazing. so ready.

**Comment by sck-at-ucy** [comment]
So exciting to see this moving along!!  We have several Studios with M2 Ultras in the lab that we could hookup to test distributed computing  when we reach that point of maturity. Would be happy to get involved in the testing.

**Comment by evilsocket** [comment]
How cool would it be being able to run 70B models by distributing them over a few ipads, iphones and whatnot? That would give new life to older devices we don't use anymore. Not to mention that it could create a paid model where users could give some of their GPU time for $ :D I am so looking forward distributed inference with MLX!

**Comment by pickettd** [comment]
I know this is an old thread. But for any of the folks above that are looking to try something out: https://github.com/exo-explore/exo supports distributed inference using either (or both) MLX and tinygrad backends (I've used it to successfully run the big Deepseek-Coder-v2 (236B parameters) across multiple Mac machines on a local network with pretty good speed on MLX)

**Comment by ivanfioravanti** [comment]
Any news on this one @awni ? The idea of running MLX distribute parallel inference on Mac connected through Thunderbolt 5 is really interesting!

**Comment by awni** [comment]
There are several PRs open for this

- Tensor parallel API https://github.com/ml-explore/mlx/pull/1270
- Tensor parallel LLMs https://github.com/ml-explore/mlx-examples/pull/890
- Tensor parallel FLUX https://github.com/ml-explore/mlx-examples/pull/1325

We haven't prioritized landing it because right now you won't see a speedup for LLMs using tensor parallelism. There is too much latency from synchronization and communication. It will take us some time to improve that because the issues there. aren't just in MLX.

The FLUX example should already run faster on multiple machines with the right setup (we'll follow up with that more once it lands).

**Comment by zcbenz** [comment]
I'm closing this issue since we already have decent distributed APIs, for specific new features let's open new issues.
